### PR TITLE
Address Ruff Rule PT021 and PIE804 violations

### PR DESCRIPTION
**Pull request by Alpha-Ursae-Minoris**
### Description

This Pull Request addresses Rule PT021and Rule PIE804.

Rule PT021 checks whether request.addfinalizer is used and suggests to "use yield instead of request.addfinalizer". 
I replaced the only use of request.addfinalizer in astropy/samp/tests/test_client.py and removed the rule in the .ruff.toml file.
Running `ruff --select=PT021 --statistics astropy/` now yields no violations.

Rule PIE804 checks for unnecessary dictionary-style keyword arguments.
I fixed the only occurrence in astropy/utils/decorators.py by explicitly passing the keyword arguments instead of unpacking them from a dictionary.  
Running `ruff --select=PIE804 --statistics astropy/` now yields no violations.

This pull request is to address Issue #14818

### CODE REVIEW

**Review by pllim — DISMISSED** [review]
I don't think your fix is working:

`astropy\samp\tests\test_client.py::test_reconnect - astropy.samp.errors.SAMPWarning: Another SAMP Hub is already running. Start-up process aborted.`

**Comment by nstarman** [comment]
Thanks @Alpha-Ursae-Minoris for the PR.
Check out https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#yield-fixtures-recommended, I think that indicates that there needs to be cleanup post-yield.

**Comment by pllim** [comment]
Also PT021 is not listed under #14818 check list.

**Comment by Alpha-Ursae-Minoris** [comment]
Thank you for your the helpful link @nstarman. Now all the tests in astropy/samp/ run successfully locally using -rs.

**Comment by Alpha-Ursae-Minoris** [comment]
@pllim I have now added PIE804 from the #14818 check list.

**Comment by pllim** [comment]
Thanks! A few things:

* Please also update ruff file with PIE804 stuff
* Please fix pre-commit failure
* Please do not create merge commit in your PR history, instead use `rebase` (let us know if you need help with this)
* Please squash all your commits into one commit

**Comment by Alpha-Ursae-Minoris** [comment]
Thanks for the pointers @pllim, they should be implemented correctly now. Please take a look and let me know if there's anything else that needs attention.

**Comment by pllim** [comment]
Thanks! mpldev failure is unrelated. The diff LGTM now but I'll let @nstarman or @eerovaher approve, since they know about Ruff way more than I do.

File: astropy/samp/tests/test_client.py [hunk 1]
```diff
@@ -35,7 +35,8 @@ def samp_hub(request):
     """A fixture that can be used by client tests that require a HUB."""
     my_hub = SAMPHubServer()
     my_hub.start()
-    request.addfinalizer(my_hub.stop)
```
Comment (eerovaher) [inline astropy/samp/tests/test_client.py:38]: This was the only line that used `request`, so `samp_hub` shouldn't be asking for it anymore.

File: astropy/samp/tests/test_client.py [hunk 2]
```diff
@@ -35,7 +35,8 @@ def samp_hub(request):
...
+    yield my_hub
```
Comment (eerovaher) [inline astropy/samp/tests/test_client.py:38]: This fixture didn't return `my_hub`, so it shouldn't yield it either.
```suggestion
    yield
```

**Review by eerovaher — CHANGES_REQUESTED** [review]
This pull request improves `astropy` by ensuring it complies with two more Ruff rules, but differently from @pllim I think there should be separate commits for both rules.

**Comment by Alpha-Ursae-Minoris** [comment]
@eerovaher I have modified the code and commits as you suggested.

**Review by eerovaher — CHANGES_REQUESTED** [review]
This is almost ready for merging, but the commit concerning `PT021` should change the line https://github.com/astropy/astropy/blob/001dfa89a03e9d2c6cdc5bff4f90dee69f1d9060/astropy/samp/tests/test_client.py#L34 to `def samp_hub():`
(The line in question is far enough from the code changes that GitHub UI doesn't let me write a code suggestion).

**Comment by Alpha-Ursae-Minoris** [comment]
@eerovaher Apologies for the oversight, should be fixed now.

**Review by eerovaher — APPROVED** [review]
Despite being a small pull request in terms of the number of edited lines, it nonetheless manages to improve `astropy` by addressing two Ruff rules.

**Pull request merged**